provd/internal/services/pro/pro.go

```diff
@@ -0,0 +1,295 @@
+				return status.Errorf(codes.Internal, fmt.Sprintf("failed to send user code response: %v", err))
+			}
+			continue
+		}
+		// Check if it was a connectivity error
+		if response.Errors.ContainsCode("connectivity-error") {
+			resp := &pb.ProMagicAttachResponse{
+				Type: pb.ProMagicAttachResponseType_NETWORK_ERROR,
+			}
+			if err := stream.Send(resp); err != nil {
+				return status.Errorf(codes.Internal, fmt.Sprintf("failed to send connectivity error response: %v", err))
+			}
+			return nil
+		}
+
+		// If not a connectivity error, return unknown error
+		resp := &pb.ProMagicAttachResponse{
+			Type: pb.ProMagicAttachResponseType_UNKNOWN_ERROR,
+		}
+		if err := stream.Send(resp); err != nil {
+			return status.Errorf(codes.Internal, fmt.Sprintf("failed to send unknown error response: %v", err))
+		}
+		return nil
+	}
+
+	// Get the contract token
+	if contractToken == nil {
+		return status.Errorf(codes.Internal, "contract token not found in response")
+	}
+
+	// Pro attach the token
+	if err := s.proExecutable.Attach(stream.Context(), *contractToken); err != nil {
+		resp := &pb.ProMagicAttachResponse{
+			Type: pb.ProMagicAttachResponseType_UNKNOWN_ERROR,
+		}
+		if err := stream.Send(resp); err != nil {
+			return status.Errorf(codes.Internal, fmt.Sprintf("failed to send unknown error response: %v", err))
+		}
+		return nil
+	}
+
+	// Send the final success response
+	successResponse := &pb.ProMagicAttachResponse{
+		Type: pb.ProMagicAttachResponseType_SUCCESS,
+	}
+	if err := stream.Send(successResponse); err != nil {
+		return status.Errorf(codes.Internal, fmt.Sprintf("failed to send final success response: %v", err))
+	}
+
+	// Close the stream
+	return nil
+}
+
+func (p *proExecutable) Initiate(ctx context.Context) (*proAPIResponse, error) {
+	out, err := exec.CommandContext(ctx, "pro", "api", "u.pro.attach.magic.initiate.v1").Output()
+	if err != nil {
+		slog.Error(fmt.Sprintf("failed to initiate magic attach: %v\nOutput: %s", err, string(out)))
+		return nil, fmt.Errorf("failed to initiate magic attach: %v\nOutput: %s", err, string(out))
+	}
+
+	// Parse the response
+	var response proAPIResponse
+	if err := json.Unmarshal(out, &response); err != nil {
+		slog.Error(fmt.Sprintf("failed to parse response: %v", err))
+		return nil, fmt.Errorf("failed to parse response: %v", err)
+	}
+
+	return &response, nil
+}
+
+func (p *proExecutable) Wait(ctx context.Context, token string) (*proAPIResponse, error) {
+	// Initiate magic attach process
+	// #nosec:G204 // We are in control of the token formatting and this is only the argument.
+	out, err := exec.CommandContext(ctx, "pro", "api", "u.pro.attach.magic.wait.v1", "--args", fmt.Sprintf("magic_token=%s", token)).Output()
+	if err != nil {
+		slog.Error(fmt.Sprintf("failed to wait on attach response: %v\nOutput: %s", err, string(out)))
+		return nil, fmt.Errorf("failed to wait on attach response: %v\nOutput: %s", err, string(out))
+	}
```

Comment:
When the client closes the stream this context gets cancelled and the error received here is `signal: killed`. That case probably shouldn't be treated as an error, since the client can close the connection for many reasons (refreshing, navigating elsewhere, etc).